grpc-protocol/src/test/java/com/linecorp/armeria/common/grpc/protocol/Base64DecoderTest.java

```diff
@@ -19,63 +19,65 @@
+        final Base64Decoder base64Decoder = new Base64Decoder(PooledByteBufAllocator.DEFAULT);
+        final ByteBuf decoded = base64Decoder.decode(Unpooled.wrappedBuffer(encoded.getBytes()));
+        assertThat(decoded.toString(Charset.defaultCharset())).isEqualTo(expected);
+        decoded.release();
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(EncodedStringProvider.class)
+    void decodeConcatenatedBufs(String expected, String encoded) {
+        final ByteBuf buf1 = Unpooled.wrappedBuffer(encoded.getBytes());
+        final ByteBuf buf2 = buf1.retainedDuplicate();
 
-    @Test
-    void decodeConcatenatedBufsWithPadding() {
-        final String str = "abcd"; // YWJjZA==
-        final ByteBuf buf = Unpooled.wrappedBuffer(str.getBytes());
-        final ByteBuf encoded1 = Base64.encode(buf);
-        buf.readerIndex(0);
-        final ByteBuf encoded2 = Base64.encode(buf);
-        final ByteBuf concatenated = Unpooled.wrappedBuffer(encoded1, encoded2); // YWJjZA==YWJjZA==
         final Base64Decoder base64Decoder = new Base64Decoder(PooledByteBufAllocator.DEFAULT);
-        final ByteBuf decoded = base64Decoder.decode(concatenated);
-        assertThat(decoded.toString(Charset.defaultCharset())).isEqualTo("abcdabcd");
+        final ByteBuf decoded = base64Decoder.decode(Unpooled.wrappedBuffer(buf1, buf2));
+        assertThat(decoded.toString(Charset.defaultCharset())).isEqualTo(expected + expected);
         decoded.release();
     }
 
-    @Test
-    void decodeMultipleEncodedBytes() {
-        final String str = "abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz1234567890";
-        final byte[] bytes = str.getBytes();
-        final List<ByteBuf> fragments = fragmentRandomly(bytes);
-        final int half = fragments.size() / 2;
-        final ByteBuf first =
-                Unpooled.wrappedBuffer(fragments.subList(0, half).toArray(EMPTY_BYTE_BUF));
-        final ByteBuf second =
-                Unpooled.wrappedBuffer(fragments.subList(half, fragments.size()).toArray(EMPTY_BYTE_BUF));
+    @ParameterizedTest
+    @ArgumentsSource(EncodedStringProvider.class)
+    void decodeEachByteSeparately(String expected, String encoded) {
+        final ByteBuf buf = Unpooled.wrappedBuffer(encoded.getBytes());
         final Base64Decoder base64Decoder = new Base64Decoder(PooledByteBufAllocator.DEFAULT);
-        final ByteBuf decodedFirst = base64Decoder.decode(first);
-        final ByteBuf decodedSecond = base64Decoder.decode(second);
-        assertThat(Unpooled.wrappedBuffer(decodedFirst, decodedSecond).toString(Charset.defaultCharset()))
-                .isEqualTo(str);
-        decodedFirst.release();
-        decodedSecond.release();
+        final int readableBytes = buf.readableBytes();
+        final List<ByteBuf> bufs = new ArrayList<>();
+        for (int i = 0; i < readableBytes; i++) {
+            bufs.add(base64Decoder.decode(buf.retainedSlice(i, 1)));
+        }
```

Comment:
Could be simplified into:

```java
while (buf.isReadable()) {
    bufs.add(base64Decoder.decode(buf.readRetainedSlice(1)));
}
```

Reply:
Fixed. 😉  Didn't know that `ByteBuf` has this API. 😄